File: vector.py
from math import sin
class Vector:

   def __init__(self,x, y, z):
      self.x = x
      self.y = y
      self.z = z

   def getMagnitude(self):
      return (self.x**2 + self.y ** 2  + self.z ** 2)**0.5

   def getMag2(self):
      return self.x**2 + self.y ** 2  + self.z ** 2     

   def setScale(self, scale):
      self.x *= scale
      self.y *= scale
      self.z *= scale
      return self

   def norm(self):
      mag=self.getMagnitude()
      self.x /= mag
      self.y /= mag
      self.z /= mag
      return self
   

   def __add__(self,v1):
      vsum = Vector(self.x + v1.x, self.y + v1.y, self.z + v1.z)
      return vsum

   def __sub__(self,v1):
      vdiff = Vector(self.x - v1.x, self.y - v1.y, self.z - v1.z)
      return vdiff

   def __mul__(self,v1):
      return self.x*v1.x + self.y*v1.y + self.z*v1.z

   def __str__(self):
      return "x: " + str(self.x) + " y: " + str(self.y)  + " z: " + str(self.z)

def crossp(v1, v2):
   cross.x = v1.y*v2.z - v1.z*v2.y
   cross.y = v1.z*v2.x - v1.x*v2.z
   cross.z = v1.x*v2.y - v1.y*v2.x
   return cross

File: Vector.java
public class Vector {
    public double x,y,z;

    public Vector(double x,double y,double z) {
        this.x = x;
        this.y = y;
        this.z = z;
    }

    public double getMagnitude() {
        return Math.sqrt(x*x + y*y + z*z);
    }

    public Vector scale(double i) {
        return new Vector(x*i, y*i, z*i);
    }

    public static Vector add(Vector v1, Vector v2) {
        return new Vector(v1.x + v2.x, v1.y + v2.y, v1.z + v2.z); 
    }

    public static Vector sub(Vector v1, Vector v2) {
        return new Vector(v1.x - v2.x, v1.y - v2.y, v1.z - v2.z); 
    }

    public Vector norm() {
        double len = this.getMagnitude();
        return this.scale(1/len);
    }

    public static Vector cross(Vector v1,Vector v2) {
        return new Vector(
                v1.y*v2.z - v1.z*v2.y,
                v1.z*v2.x - v1.x*v2.z,
                v1.x*v2.y - v1.y*v2.x);
    }

    public static double dot(Vector v1,Vector v2) {
        return v1.x * v2.x + v1.y * v2.y + v1.z * v2.z; 
    }

    public Vector neg() {
        return new Vector(
                -this.x,
                -this.y,
                -this.z);
    }

    @Override
    public String toString() {
        return Double.toString(this.x) + " "  + 
               Double.toString(this.y) + " "  + 
               Double.toString(this.z);
    }
}

